pkg/sql/catalog/descs/collection.go

```diff
@@ -308,16 +308,42 @@ func (tc *Collection) WriteDesc(
 	return txn.Run(ctx, b)
 }
 
-// GetDescriptorsWithNewVersion returns all the IDVersion pairs that have
-// undergone a schema change. Returns nil for no schema changes. The version
-// returned for each schema change is clusterVersion - 1, because that's the one
-// that will be used when checking for table descriptor two version invariance.
-func (tc *Collection) GetDescriptorsWithNewVersion() (originalVersions []lease.IDVersion) {
-	_ = tc.uncommitted.iterateNewVersionByID(func(originalVersion lease.IDVersion) error {
-		originalVersions = append(originalVersions, originalVersion)
+// GetOriginalPreviousIDVersionsForUncommitted returns all the IDVersion
+// pairs for descriptors that have undergone a schema change.
+// Returns an empty slice for no schema changes.
+//
+// The version returned for each schema change is clusterVersion - 1, because
+// that's the one that will be used when checking for table descriptor
+// two-version invariance.
+func (tc *Collection) GetOriginalPreviousIDVersionsForUncommitted() (
+	withNewVersions []lease.IDVersion,
+	err error,
+) {
+	err = tc.uncommitted.iterateOriginalByID(func(original catalog.Descriptor) error {
+		uncommitted := tc.uncommitted.getUncommittedByID(original.GetID())
+		// Sanity check.
+		if original.GetVersion() == 0 {
+			return errors.AssertionFailedf(
+				"expected original version of uncommitted %s %q (%d) to be non-zero",
+				uncommitted.DescriptorType(), uncommitted.GetName(), uncommitted.GetID())
+		}
+		// Ignore uncommitted descriptors which for whatever reason have not
```

Comment:
Do you understand the reason?

Reply:
For a reason unknown to me we allow uncommitted descriptors to be added with the same version as the cluster version, in addition to clusterVersion+1. These were previously excluded by an early return on `mut.IsUncommittedDescriptor()`.

Reply:
Perhaps we should file an issue to tighten this. I can't see any instances where `AddUncommittedDescriptor` isn't preceded by a `MaybeIncrementVersion` call, unless it's obviously a new descriptor.

Reply:
In the old world, before your rewrite, we'd add to `uncommittedDescriptors` or whatever it was then called (`storedDescriptors`) on the read path, and the caller which resolved that descriptor might decide to throw that away. 

I'd still like to understand how we were modifying this set which, as you say, shouldn't have anything added to it unless it is being written. One note is that the thing which triggers the bug is the underlying call to `GetAllDescriptors`.

Reply:
I looked into this and the bug is deeper, even though the changes brought by this PR are good. 

The problem is that we upsert into the storage cache when really we should be inserting. Consider this test case which you supplied. What happens in that transaction is that `CREATE SCHEMA sc2` adds an uncommitted db descriptor to the collection with the new schema entry and bumps its version counter. It then `Run`s that batch. The subsequent `crdb_internal.tables` query calls a `GetAllDescriptors` which overwrites the storage cache with the uncommitted descriptor!

This is why the `original` descriptors in the uncommitted layer can't be relied on.

Reply:
>This is why the original descriptors in the uncommitted layer can't be relied on. 

Yeah... this seems like the right answer. I think we should avoid touching `original` in `AddUncommittedDescriptor` when we already have an entry for the descriptor as opposed to pulling from the `stored` cache.

Reply:
Indeed. I'll revert to draft mode and push a fix in that direction and we'll see what happens.